src/cutlass_mlp.cu

```diff
@@ -314,6 +327,606 @@ void CutlassMLP<T>::backward_impl(
+template <typename T>
+__global__ void element_wise_copy_CM_RM(uint32_t n_elements, uint32_t row, uint32_t col, T* tmp, T* res) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	uint32_t idx_i = i % row;
+	uint32_t idx_j = i / row;
+	uint32_t i_RM = idx_i * col + idx_j;
+	res[i_RM] = tmp[i];
+}
+
+// compute d_doutputdp2_dinput
+template <typename T>
+__global__ void multiply_w_RM(uint32_t n_elements, uint32_t w_row, uint32_t w_col, T* weights, T* doutputdp_2, T* result) {
+	uint32_t i = blockIdx.x * blockDim.x + threadIdx.x;
+	if (i >= n_elements) return;
+
+	uint32_t idx_doutputdp = i % w_row;
+	T tmp = doutputdp_2[idx_doutputdp];
+	uint32_t i_RM = (i % w_row) * w_col + i / w_row;
+	result[i] = weights[i_RM] * tmp;
+}
+
+// compute dL2_ddL1doutput: activation Softplus/ReLU derivative, doutputdp and then multiply dL2_ddL1doutput_tmp
+template <typename CutlassLayer, typename T>
+bool compute_dL2_ddL1doutput(
+	cudaStream_t stream,
+	bool is_inference,
+	Activation activation,
+	GPUMatrix<T, RM>& weights,
+	const GPUMatrix<T, CM>& p,
+	const GPUMatrixDynamic<T>& dL_ddLdinput,
+	GPUMatrixDynamic<T>& dL2_ddL1doutput,
+	GradientMode param_gradients_mode
+) {
+	// dL2_ddL1doutput = weight x dL_ddLdinput where x is matrix multiply
+	if (weights.layout() == CM) {
+		fc_multiply<CutlassLayer>(stream, weights.cm(), dL_ddLdinput, dL2_ddL1doutput);
+	} else {
+		fc_multiply<CutlassLayer>(stream, weights.rm(), dL_ddLdinput, dL2_ddL1doutput);
+	}
+
+	// if activation is None, dL2_ddL1doutput = weight x dL2_ddL1dinput
+	if (activation == Activation::None) {
+		return true;
+	}
+
+	// dL2_ddL1doutput = dL2_ddL1doutput · doutputdp, where p is linear output, · is dot product
+	activation_backward_gpu(stream, activation, p, dL2_ddL1doutput);
+	
+	return true;
+}
+
+template <typename CutlassLayer, typename T>
+bool compute_dL2dw(
+	cudaStream_t stream,
+	bool is_inference,
+	Activation activation,
+	GPUMatrix<T, RM>& weights,
+	const GPUMatrixDynamic<T>& input,
+	const GPUMatrixDynamic<T>& p, // const GPUMatrix<T, CM>& p
+	const GPUMatrixDynamic<T>& dL_dp,
+	const GPUMatrixDynamic<T>& dL_doutput,
+	const GPUMatrixDynamic<T>& dL_ddLdinput, // dL2_ddL1dinput
+	const GPUMatrixDynamic<T>& ddoutputdp_dp, // ddoutputdp_dp
+	GPUMatrixDynamic<T>& dL2_ddL1doutput, // pointer better
+	GPUMatrix<T, RM>& weight_gradient, // pointer better
+	GradientMode param_gradients_mode
+) {
+	// dL2dw_1 = ddL1dinput_dw x dL2_ddL1dinput	
+	uint32_t batch_size = dL_ddLdinput.n();
+	int split_k_factor = batch_size / std::min((uint32_t)(1 << 12), batch_size);
+	const float param_gradient_beta = param_gradients_mode == GradientMode::Accumulate ? 1.0f : 0.0f;
+
+	//cudaStream_t stream_dL2dw_1;
+	if (param_gradients_mode != GradientMode::Ignore) {
+		fc_multiply_split_k<FullLayerK>(stream, dL_dp, dL_ddLdinput.transposed(), weight_gradient, split_k_factor, param_gradient_beta);
+	}
+
+	// when activation is None, don't have to compute dL2dw_2
```

Comment:
Redundant comment. If you'd like to have a comment here, it should explain *why* dL2dw_2 does not need to be computed if the activation is none. (Though I think most readers familiar with derivatives will accept it in any case.)